pkg/completion/completion.go

```diff
@@ -67,33 +68,47 @@ func (wg *WaitGroup) AddCompletion() *Completion {
 // Wait blocks until all completions added by calling AddCompletion are
 // completed, or the context is canceled, whichever happens first.
 // Returns the context's error if it is cancelled, nil otherwise.
+// No callbacks of the completions in this wait group will be called after
+// this returns.
+// Returns the error value of one of the completions, if available, or the
+// error value of the WaitGroup otherwise.
 func (wg *WaitGroup) Wait() error {
 	wg.counterLocker.Lock()
 	defer wg.counterLocker.Unlock()
 
+	var err error
 Loop:
 	for i, comp := range wg.pendingCompletions {
 		select {
 		case <-comp.Completed():
+			err = comp.Error
 			continue Loop
 		case <-wg.ctx.Done():
 			// Complete the remaining completions to make sure their completed
 			// channels are closed.
+			wgErr := wg.ctx.Err() // context.Canceled or context.DeadlineExeeded
 			for _, comp := range wg.pendingCompletions[i:] {
-				comp.complete(false)
+				comp.Complete(wgErr)
+				// Note if the completion errored
+				if err == nil && comp.Error != wgErr {
+					err = comp.Error
+				}
+			}
+			if err == nil {
+				err = wgErr
 			}
 			break Loop
 		}
 	}
 	wg.pendingCompletions = nil
-	return wg.ctx.Err()
+	return err
 }
 
 // Completion provides the Complete callback to be called when an asynchronous
 // computation is completed.
 type Completion struct {
-	// ctx is the context of the wait group.
-	ctx context.Context
+	// wg is the wait group the completion belongs to
+	wg *WaitGroup
```

Comment:
I would prefer to avoid the circular dependency. Instead of replacing `ctx` with `wg`, I would just add the `cancel` function, since accessing `cancel` is the only reason you did this change.

Reply:
This change requires the Completion.Context() method to be removed. Doing that.